source/common/http/http1/parser_impl.cc

```diff
@@ -0,0 +1,225 @@
+          auto status = conn_impl->onMessageBegin();
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onUrl(at, length);
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        nullptr, // on_status
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onHeaderField(at, length);
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onHeaderValue(at, length);
+          return conn_impl->setAndCheckCallbackStatus(std::move(status));
+        },
+        [](llhttp_t* parser) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto statusor = conn_impl->onHeadersComplete();
+          return conn_impl->setAndCheckCallbackStatusOr(std::move(statusor));
+        },
+        [](llhttp_t* parser, const char* at, size_t length) -> int {
+          static_cast<ParserCallbacks*>(parser->data)->bufferBody(at, length);
+          return 0;
+        },
+        [](llhttp_t* parser) -> int {
+          auto* conn_impl = static_cast<ParserCallbacks*>(parser->data);
+          auto status = conn_impl->onMessageComplete();
+          return conn_impl->setAndCheckCallbackStatusOr(std::move(status));
+        },
+        [](llhttp_t* parser) -> int {
+          // A 0-byte chunk header is used to signal the end of the chunked body.
+          // When this function is called, http-parser holds the size of the chunk in
+          // parser->content_length. See
+          // https://github.com/nodejs/http-parser/blob/v2.9.3/http_parser.h#L336
+          const bool is_final_chunk = (parser->content_length == 0);
+          static_cast<ParserCallbacks*>(parser->data)->onChunkHeader(is_final_chunk);
+          return 0;
+        },
+        nullptr, // on_chunk_complete
+        nullptr, // on_url_complete
+        nullptr, // on_status_complete
+        nullptr, // on_header_field_complete
+        nullptr  // on_header_value_complete
+    };
+  }
+
+  RcVal execute(const char* slice, int len) {
+    llhttp_errno_t error;
+    if (slice == nullptr || len == 0) {
+      error = llhttp_finish(&parser_);
+    } else {
+      error = llhttp_execute(&parser_, slice, len);
+    }
+    size_t nread = len;
+    // Adjust number of bytes read in case of error.
+    if (error != HPE_OK) {
+      nread = llhttp_get_error_pos(&parser_) - slice;
+      // Resume after upgrade.
+      if (error == HPE_PAUSED_UPGRADE) {
+        error = HPE_OK;
```

Comment:
It seems like we handle upgrade by manually setting HPE_PAUSED in the codec (not ever HPE_PAUSED_UPGRADE), so this might be why we don't actually need to follow this contract
https://github.com/nodejs/llhttp/blob/674cf94e975bea73970015eae272088b571cf368/src/native/api.h#L117
@derekargueta do you have more context on this?

Reply:
Seems like we could remove it -- the codec handling upgrade is the reason why. We always directly return HPE_PAUSED when we are upgrading. I think probably we can add an ENVOY_BUG here that the error is not HPE_PAUSED_UPGRADE, since ideally we'd like to detect if this is not the case while we're canarying.